src/server.c

```diff
@@ -209,19 +209,20 @@ mstime_t mstime(void) {
  * reflect the same time.
  * More details can be found in the comments below. */
 mstime_t commandTimeSnapshot(void) {
-    /* If we are in the context of a Lua script, we pretend that time is
-     * blocked to when the Lua script started. This way a key can expire
-     * only the first time it is accessed and not in the middle of the
-     * script execution, making propagation to slaves / AOF consistent.
-     * See issue #1525 on Github for more information. */
-    /* If we are in the middle of a command execution, we still want to use
-     * a reference time that does not change: in that case we just use the
+    /* When we are in the middle of a command execution, we want to use a
+     * reference time that does not change: in that case we just use the
      * cached time, that we update before each call in the call() function.
      * This way we avoid that commands such as RPOPLPUSH or similar, that
      * may re-open the same key multiple times, can invalidate an already
      * open object in a next call, if the next call will see the key expired,
-     * while the first did not. */
-    return server.mstime;
+     * while the first did not.
+     * This is specificlally important in the context of scripts, where we
+     * pretend that time freezes. This way a key can expire only the first time
+     * it is accessed and not in the middle of the script execution, making
+     * propagation to slaves / AOF consistent. See issue #1525 for more info.
+     * Note that we cannot use the cached server.mstime because it can chagne
+     * in processEventsWhileBlocked etc. */
+    return server.cmd_time_snapshot;
```

Comment:
so if a script is timed out, all commands that are processed until the script is done will have the same time snapshot that was taken when the script started?

Reply:
the other alternative is to backup and restore it in blockingOperationStarts / Ends.
do you prefer that? @soloestoy WDYT?

Reply:
i think backup/restore is a good idea

Reply:
i implemented what's relatively easy. i'm not really happy with it (see the long comment in my last commit).
i suppose that although not perfect, we can live with that.

Reply:
uh... actually beside this, we have lots of places call `updateCachedTime`, like:
1. moduleGILAfterLock
2. afterSleep
3. blockingOperationStarts and its callers 

now I'm thinking do we need ensure time is fixed in scripts? maybe just make sure time is fixed in one command is enough? since we propagate commands instead script itself.

Reply:
@soloestoy i'm not certain i understand what you mean.
do you mean that now that scripts are propagated as effect (rather than forwarding the EVAL), there's no longer a need to freeze time during the execution of the script?
i don't think that's right. if the script reads the same key twice, it doesn't make sense that it can first exist and later be missing.

regarding the 1-3 bullets above, i don't think they cause any issue. all these places are updating **the server's** cached time (possibly after some period of inactivity), without them, it could mean that the server time is very much outdated, and with them, it's **relatively** fresh.

but for commands (including scripts and module command that could be long and yield as well), we want a more accurate time, and we want to be sure it's frozen, and i think my last commit handles that ok.

The downside of my last commit is that commands that are executed during an yielding script or module, will all have the same ("relatively" fresh) time, and not a precise one like commands outside of yielding script / module.
but considering these are normally just introspection and kill commands, i think it should be enough.